### PR TITLE
Rework socket

### PR DESCRIPTION
**Pull request by srittau**
* Extract _socket.pyi from socket.pyi.
* Extract _socket.socket from socket.socket.
* Fix socket.family annotation.
* Annotate SocketIO properly.
* SocketType is an alias of _socket.socket.
* Sort items in socket.pyi in the same order as in socket.py.
* Remove socket.EINTR.
* Add socket.errorTab (Windows only) and _socket.dup().
* Use _typeshed.WriteableBuffer instead of custom alias.
* Use built-in generics.
* Use socket.socket rather than socket.SocketType in stdlib and third-party stubs.

### CODE REVIEW

**Comment by srittau** [comment]
Ok, it seems that mypy doesn't handle overridden star imports correctly. :/

**Comment by srittau** [comment]
Looking at the primer failures:

* `anyio.abc._sockets` wasn't flagged before, since it's the `attributes` dict contained an `Any` type before.
* `anyio._backends._asyncio` wasn't flagged before, since `UNIXSocketStream.__raw_socket` was annotated as `socket.SocketType`, which used to be an alias for `Any`. The annotation is wrong and should be `socket.socket`, which should fix this.
* `paasta` and `ignite` don't seem to have changes. I don't know why a diff is generated. Cc @hauntsaninja Do you know what's going on?
* `werkzeug`: This points to a genuine problem that wasn't flagged before, since `BaseRequestHandler.connection` was `SocketType`, which was an alias for `Any`.

**Comment by srittau** [comment]
I opened [bpo-44261](https://bugs.python.org/issue44261). `SocketType` is an alias for `_socket.socket`, when it is documented to be (and probably should be) an alias for `socket.socket`, a subclass.

File: stdlib/_socket.pyi [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+    # PF_SYSTEM is defined by macOS
+    PF_SYSTEM: int
+    SYSPROTO_CONTROL: int
+
+# ----- Exceptions -----
+
+error = OSError
+
+class herror(error):
+    def __init__(self, herror: int = ..., string: str = ...) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:508]:
```
In [14]: _socket.herror(herror=42)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-14-6d338928f131> in <module>
----> 1 _socket.herror(herror=42)

TypeError: herror() takes no keyword arguments
```

I think the `__init__` method for this and the next two should just be removed. They just inherit from OSError.

File: stdlib/_socket.pyi [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+    def __init__(self, error: int = ..., string: str = ...) -> None: ...
+
+# ----- Classes -----
+
+class socket:
+    family: int
+    type: int
+    proto: int
+    def __init__(self, family: int = ..., type: int = ..., proto: int = ..., fileno: Optional[int] = ...) -> None: ...
+    def bind(self, address: Union[_Address, bytes]) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:523]: Missing the close method
```suggestion
    def bind(self, address: Union[_Address, bytes]) -> None: ...
    def close(self) -> None: ...
```

File: stdlib/_socket.pyi [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+    def connect(self, address: Union[_Address, bytes]) -> None: ...
+    def connect_ex(self, address: Union[_Address, bytes]) -> int: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:525]:
```suggestion
    def connect_ex(self, address: Union[_Address, bytes]) -> int: ...
    def detach(self) -> None: ...
```
Another missing one

File: stdlib/_socket.pyi (hunk 2 above)
Comment (JelleZijlstra) [reply stdlib/_socket.pyi:523]: Also, bind only takes positional args. The same is true for the other METH_O methods: bind, connect, connect_ex, setblocking, settimeout, shutdown.

File: stdlib/_socket.pyi [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+    def getsockopt(self, level: int, optname: int) -> int: ...
+    @overload
+    def getsockopt(self, level: int, optname: int, buflen: int) -> bytes: ...
+    if sys.version_info >= (3, 7):
+        def getblocking(self) -> bool: ...
+    def gettimeout(self) -> Optional[float]: ...
+    if sys.platform == "win32":
+        def ioctl(self, control: int, option: Union[int, tuple[int, int, int], bool]) -> None: ...
+    def listen(self, __backlog: int = ...) -> None: ...
+    def recv(self, bufsize: int, flags: int = ...) -> bytes: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:539]:
```suggestion
    def recv(self, __bufsize: int, __flags: int = ...) -> bytes: ...
```
This is METH_VARARGS so it doesn't take keywords

File: stdlib/_socket.pyi [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+    def recvfrom(self, bufsize: int, flags: int = ...) -> tuple[bytes, _RetAddress]: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:540]:
```suggestion
    def recvfrom(self, __bufsize: int, __flags: int = ...) -> tuple[bytes, _RetAddress]: ...
```
Also METH_VARARGS

File: stdlib/_socket.pyi [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+    if sys.platform != "win32":
+        def recvmsg(self, __bufsize: int, __ancbufsize: int = ..., __flags: int = ...) -> tuple[bytes, list[_CMSG], int, Any]: ...
+        def recvmsg_into(
+            self, __buffers: Iterable[WriteableBuffer], __ancbufsize: int = ..., __flags: int = ...
+        ) -> tuple[int, list[_CMSG], int, Any]: ...
+    def recvfrom_into(self, buffer: WriteableBuffer, nbytes: int = ..., flags: int = ...) -> tuple[int, _RetAddress]: ...
+    def recv_into(self, buffer: WriteableBuffer, nbytes: int = ..., flags: int = ...) -> int: ...
+    def send(self, data: bytes, flags: int = ...) -> int: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:548]: send, sendall, sendto are also METH_VARARGS so they only take positional args

File: stdlib/_socket.pyi [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+            self, msg: Iterable[bytes] = ..., *, op: int, iv: Any = ..., assoclen: int = ..., flags: int = ...
+        ) -> int: ...
+    def setblocking(self, flag: bool) -> None: ...
+    def settimeout(self, value: Optional[float]) -> None: ...
+    @overload
+    def setsockopt(self, level: int, optname: int, value: Union[int, bytes]) -> None: ...
+    @overload
+    def setsockopt(self, level: int, optname: int, value: None, optlen: int) -> None: ...
+    if sys.platform == "win32":
+        def share(self, process_id: int) -> bytes: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:569]: Another METH_VARARGS
```suggestion
        def share(self, __process_id: int) -> bytes: ...
```

File: stdlib/_socket.pyi [hunk 8]
```diff
@@ -0,0 +1,619 @@
...
+    def shutdown(self, how: int) -> None: ...
+
+SocketType = socket
+
+# ----- Functions -----
+
+if sys.version_info >= (3, 7):
+    def close(fd: int) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:577]:
```suggestion
    def close(__fd: int) -> None: ...
```
METH_O

Comment (JelleZijlstra) [reply stdlib/_socket.pyi:577]: Also, this should accept `SupportsIndex`. Probably good to make a type alias for that since this affects all functions here that take an fd argument and pass it to `PyLong_AsSocket_t`.

File: stdlib/_socket.pyi [hunk 9]
```diff
@@ -0,0 +1,619 @@
...
+
+def dup(__fd: int) -> int: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:579]: SupportsIndex

File: stdlib/_socket.pyi [hunk 10]
```diff
@@ -0,0 +1,619 @@
...
+
+# the 5th tuple item is an address
+def getaddrinfo(
+    host: Optional[Union[bytearray, bytes, str]],
+    port: Union[str, int, None],
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:584]:
```suggestion
    port: Union[bytes, str, int, None],
```
There's a `PyBytes_Check` allowing bytes.

Also, a bit confusing that this one uses `Union[... None]` and the host arg uses `Optional`.

File: stdlib/_socket.pyi [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+    family: int = ...,
+    type: int = ...,
+    proto: int = ...,
+    flags: int = ...,
+) -> list[tuple[int, int, int, str, Union[tuple[str, int], tuple[str, int, int, int]]]]: ...
+def gethostbyname(hostname: str) -> str: ...
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:590]: This is METH_VARARGS so it doesn't take keyword args. Looks like all global functions except getaddrinfo only take positional args.

File: stdlib/socket.pyi [hunk 12]
```diff
@@ -1,509 +1,385 @@
...
 
-# ----- Constants -----
-# Some socket families are listed in the "Socket families" section of the docs,
-# but not the "Constants" section. These are listed at the end of the list of
-# constants.
-#
-# Besides those and the first few constants listed, the constants are listed in
-# documentation order.
+import _socket
+from _socket import (
```
Comment (JelleZijlstra) [inline stdlib/socket.pyi:10]: Why can't we just do `from _socket import *`? That's what happens at runtime, and it's easier to maintain than keeping that huge list of constants in sync.

File: stdlib/socket.pyi [hunk 13]
```diff
@@ -1,509 +1,385 @@
...
+_T = TypeVar("_T")
+
+# Re-exported from errno
+EBADF: int
+EAGAIN: int
+EWOULDBLOCK: int
 
-# enum versions of above flags
 class AddressFamily(IntEnum):
     AF_UNIX: int
```
Comment (JelleZijlstra) [inline stdlib/socket.pyi:384]: This technically needs some more platform dependence. On Mac I only get these:
```
In [29]: socket.AddressFamily.__members__
Out[29]: 
mappingproxy({'AF_UNSPEC': <AddressFamily.AF_UNSPEC: 0>,
              'AF_UNIX': <AddressFamily.AF_UNIX: 1>,
              'AF_INET': <AddressFamily.AF_INET: 2>,
              'AF_SNA': <AddressFamily.AF_SNA: 11>,
              'AF_APPLETALK': <AddressFamily.AF_APPLETALK: 16>,
              'AF_ROUTE': <AddressFamily.AF_ROUTE: 17>,
              'AF_LINK': <AddressFamily.AF_LINK: 18>,
              'AF_IPX': <AddressFamily.AF_IPX: 23>,
              'AF_INET6': <AddressFamily.AF_INET6: 30>,
              'AF_SYSTEM': <AddressFamily.AF_SYSTEM: 32>})
```
I'm fine with leaving this as is though.

File: stdlib/socket.pyi [hunk 14]
```diff
@@ -648,62 +567,58 @@ class socket:
...
-    @overload
-    def setsockopt(self, level: int, optname: int, value: Union[int, bytes]) -> None: ...
-    @overload
-    def setsockopt(self, level: int, optname: int, value: None, optlen: int) -> None: ...
-    if sys.platform == "win32":
-        def share(self, process_id: int) -> bytes: ...
-    def shutdown(self, how: int) -> None: ...
 
-# ----- Functions -----
+def fromfd(fd: int, family: Union[AddressFamily, int], type: Union[SocketKind, int], proto: int = ...) -> socket: ...
```
Comment (JelleZijlstra) [inline stdlib/socket.pyi:580]: fd can be SupportsIndex

File: tests/stubtest_whitelists/py3_common.txt [hunk 15]
```diff
@@ -12,6 +12,7 @@ _collections_abc.Mapping.get  # Adding None to the Union messed up mypy
 _collections_abc.Sequence.index  # Supporting None in end is not mandatory
 _csv.Dialect.__init__  # C __init__ signature is inaccurate
 _dummy_threading
+_socket.*
```
Comment (JelleZijlstra) [inline tests/stubtest_whitelists/py3_common.txt:15]: Why is this needed?

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Thanks, noticed some improvements in the C stubs

File: stdlib/_socket.pyi (hunk 1 above)
Comment (srittau) [reply stdlib/_socket.pyi:508]: And it seems that `_socket.timeout` is an alias for `TimeoutError`, starting with Python 3.10. Will change.

File: stdlib/_socket.pyi (hunk 2 above)
Comment (srittau) [reply stdlib/_socket.pyi:523]: Also a few functions and everything marked with `METH_VARARGS`. Changing.

File: stdlib/_socket.pyi (hunk 8 above)
Comment (srittau) [reply stdlib/_socket.pyi:577]: Affects `dup()`, top-level `close()`, and the `socket()` constructor. Changing.

File: stdlib/_socket.pyi (hunk 10 above)
Comment (srittau) [reply stdlib/_socket.pyi:584]: They are now consistently using PEP 604 syntax. 🎉 Also, I removed the redundant `bytearray` from `host`.

File: stdlib/socket.pyi (hunk 12 above)
Comment (srittau) [reply stdlib/socket.pyi:10]: I tried this at first. See my frustrated comment about mypy not handling overrides of star imports correctly above.

File: stdlib/socket.pyi (hunk 13 above)
Comment (srittau) [reply stdlib/socket.pyi:384]: I suspected that these are not up-to-date anymore. But I'm not keen on looking through this in detail at the moment.

File: tests/stubtest_whitelists/py3_common.txt (hunk 15 above)
Comment (srittau) [reply tests/stubtest_whitelists/py3_common.txt:15]: See above for the problems with the star imports. Since the star imports are now gone, this isn't needed anymore. Removing.

File: stdlib/socket.pyi (hunk 12 above)
Comment (srittau) [reply stdlib/socket.pyi:13]: I added a comment to the import that explains this decision.

File: stdlib/_socket.pyi [hunk 16]
```diff
@@ -0,0 +1,627 @@
+import sys
+from _typeshed import WriteableBuffer
+from collections.abc import Iterable
+from typing import Any, Optional, SupportsInt, Tuple, Union, overload
+
+if sys.version_info >= (3, 8):
```
Comment (JelleZijlstra) [inline stdlib/_socket.pyi:6]: I was a bit surprised by this but it seems right; PyLong_AsSocket_t is an alias for PyLong_AsLong, and that does `__index__` on main but not on 3.7.

Comment (srittau) [reply stdlib/_socket.pyi:6]: I'd like to claim that was intentional, but it was pure lazyness on my part, since I didn't want to write a custom protocol. Of course, I could have imported from typing_extensions ...

**Review by JelleZijlstra — APPROVED** [review]
